geo/src/algorithm/relate/geomgraph/index/rstar_edge_set_intersector.rs

```diff
@@ -1,105 +1,64 @@
-use super::super::Edge;
-use super::{EdgeSetIntersector, SegmentIntersector};
+use super::super::{Edge, GeometryGraph};
+use super::{EdgeSetIntersector, Segment, SegmentIntersector};
 use crate::GeoFloat;
 
 use std::cell::RefCell;
 use std::rc::Rc;
 
-use rstar::RTree;
+use rstar::{RTree, RTreeNum};
 
-pub(crate) struct RstarEdgeSetIntersector;
+pub(crate) struct RStarEdgeSetIntersector;
 
-impl RstarEdgeSetIntersector {
-    pub fn new() -> Self {
-        RstarEdgeSetIntersector
-    }
-}
-
-struct Segment<'a, F: GeoFloat + rstar::RTreeNum> {
-    i: usize,
-    edge: &'a RefCell<Edge<F>>,
-    envelope: rstar::AABB<crate::Coord<F>>,
-}
-
-impl<'a, F> Segment<'a, F>
-where
-    F: GeoFloat + rstar::RTreeNum,
-{
-    fn new(i: usize, edge: &'a RefCell<Edge<F>>) -> Self {
-        use rstar::RTreeObject;
-        let p1 = edge.borrow().coords()[i];
-        let p2 = edge.borrow().coords()[i + 1];
-        Self {
-            i,
-            edge,
-            envelope: rstar::AABB::from_corners(p1, p2),
-        }
-    }
-}
-
-impl<'a, F> rstar::RTreeObject for Segment<'a, F>
+impl<F> EdgeSetIntersector<F> for RStarEdgeSetIntersector
 where
-    F: GeoFloat + rstar::RTreeNum,
-{
-    type Envelope = rstar::AABB<crate::Coord<F>>;
-
-    fn envelope(&self) -> Self::Envelope {
-        self.envelope
-    }
-}
-
-impl<F> EdgeSetIntersector<F> for RstarEdgeSetIntersector
-where
-    F: GeoFloat + rstar::RTreeNum,
+    F: GeoFloat + RTreeNum,
 {
     fn compute_intersections_within_set(
-        &mut self,
-        edges: &[Rc<RefCell<Edge<F>>>],
+        &self,
+        graph: &mut GeometryGraph<F>,
         check_for_self_intersecting_edges: bool,
         segment_intersector: &mut SegmentIntersector<F>,
     ) {
-        let segments: Vec<Segment<F>> = edges
-            .iter()
-            .flat_map(|edge| {
-                let start_of_final_segment: usize = RefCell::borrow(edge).coords().len() - 1;
-                (0..start_of_final_segment).map(|segment_i| Segment::new(segment_i, edge))
-            })
-            .collect();
-        let tree = RTree::bulk_load(segments);
-
-        for (edge0, edge1) in tree.intersection_candidates_with_other_tree(&tree) {
-            if check_for_self_intersecting_edges || edge0.edge.as_ptr() != edge1.edge.as_ptr() {
```

Comment:
Sending two mutable references to `add_intersection` doesn't work without the Rc<RefCell<Edge>>. But this results in intersections being added to the same edge twice. Is that intentional?

Reply:
> But this results in intersections being added to the same edge twice. Is that intentional?

I'm not sure I understand what you mean, but it sounds like a bug. Can you provide a little example?

Reply:
Here: https://github.com/gauteh/geo/blob/graph-send/geo/src/algorithm/relate/geomgraph/index/segment_intersector.rs#L139 both edge0 and edge1 are pointing to the same object for the case when `check_for_self_intersecting_edges` is `true`.

Reply:
Ah thanks for the link. Based on the comment it seems like it wasn't an accident, but it seems weird, and to be honest, I don't remember why we'd want that.

It seems to be present in the JTS source that this was ported from. That means it's probably, but not necessarily correct 😄:

https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/geomgraph/index/SegmentIntersector.java#L179

It might be a deeper dive than I have time for today to answer this question, but if you'd like to dig deeper and document any findings that could be helpful. i.e. do all the tests still pass if we "fix" it?

Reply:
I will probably just have to reproduce the behaviour for now, can put in a comment for the future.

Reply:
Made separate methods for testing against self.